Services/Utilities/classes/class.ilUtil.php

```diff
@@ -3632,8 +3628,25 @@ public static function redirect($a_script)
         // is never written, which is a nightmare to develop with.
         session_write_close();
 
-		header("Location: ".$a_script);
-		exit();
+		global $DIC;
+		$http = $DIC->http();
+		switch (true) {
+			case ($http->request()->getQueryParams()[ilFileStandardDropzoneInputGUI::ASYNC_FILEUPLOAD]):
+				$body = new ILIAS\Filesystem\Stream\Stream(fopen('data://text/plain,' . json_encode([
+						'success'      => true,
+						'message'      => 'Called redirect after async fileupload request',
+						"redirect_url" => $a_script,
+					]), 'r'));
```

Comment:
Again another reason why ilCtrl might be a good place to handle this, since redirects are usually triggered through ilCtrl. Even creating a dependency on ilCtrl in this ilUtil function would be ok for now imo.